scipy/stats/_distn_infrastructure.py

```diff
@@ -3855,3 +3873,437 @@ def get_distribution_names(namespace_pairs, rv_base_class):
+
+    """
+
+    def __init__(self, dist, data, discrete, nllf, res):
+        self.dist = dist
+        self.data = data
+        self.discrete = discrete
+        self.pxf = getattr(dist, "pmf", None) or getattr(dist, "pdf", None)
+        self.success = getattr(res, "success", None)
+        self.message = getattr(res, "message", None)
+
+        shape_names = [] if dist.shapes is None else dist.shapes.split(", ")
+        if not discrete:
+            FitParams = namedtuple('FitParams', shape_names + ['loc', 'scale'])
+        else:
+            FitParams = namedtuple('FitParams', shape_names + ['loc'])
+
+        self.params = FitParams(*res.x)
+        self.objective_val = getattr(res, "fun", None)
+        if self.objective_val is None:
+            self.objective_val = nllf(res.x)
+        self.nllf = nllf
+
+    def __repr__(self):
+        keys = ["dist", "data", "nllf", "params", "objective_val",
+                "success", "message"]
+        m = max(map(len, keys)) + 1
+        return '\n'.join([key.rjust(m) + ': ' + repr(getattr(self, key))
+                          for key in keys if getattr(self, key) is not None])
+
+    def plot(self, ax=None):
+        """Visualize the fit result.
+
+        Superposes the PDF/PMF of the fitted distribution over a normalized
+        histogram of the data.
+
+        Available only if ``matplotlib`` is installed.
+
+        Parameters
+        ----------
+        ax : matplotlib.axes.Axes
+            Axes object to draw the plot onto, otherwise uses the current Axes.
+
+        Returns
+        -------
+        ax : matplotlib.axes.Axes
+            The matplotlib Axes object on which the plot was drawn.
+        """
+        try:
+            from matplotlib.ticker import MaxNLocator  # type: ignore[import]
+        except ModuleNotFoundError:
+            message = "matplotlib must be installed to use method `plot`."
+            raise ValueError(message)
+
+        if ax is None:
+            import matplotlib.pyplot as plt  # type: ignore[import]
+            ax = plt.gca()
+
+        fit_params = np.atleast_1d(self.params)
+        support = self.dist.support(*fit_params)
+        lb = support[0] if np.isfinite(support[0]) else min(self.data)
+        ub = support[1] if np.isfinite(support[1]) else max(self.data)
+
+        if self.discrete:
+            x = np.arange(lb, ub + 2)
+            y = self.pxf(x, *fit_params)
+            ax.vlines(x[:-1], 0, y[:-1], label='Fit Distribution PMF',
+                      color='C0')
+            options = dict(density=True, bins=x, align='left', color='C1')
+            ax.xaxis.set_major_locator(MaxNLocator(integer=True))
+
+        else:
+            x = np.linspace(lb, ub, 200)
+            y = self.pxf(x, *fit_params)
+            ax.plot(x, y, '--', label='Fit Distribution PDF', color='C0')
+            options = dict(density=True, bins=50, align='mid', color='C1')
+
+        ax.hist(self.data, label="Histogram of Data", **options)
```

Comment:
I would remove the histogram. As discussed a bit [here](https://github.com/tupui/scipy-magic/issues/1), it does not really add information. What I would prefer to see would be discrete samples bellow the PDF (would not render properly for PMF I think), cf the linked repo for instance.

Reply:
I agree that in that example, "it gives strictly less information than the PDF curve". That's not true here. The PDF was fit to the data, so it makes sense to see some representation of the original data.

This is the first of four goodness-of-fit plots provided by the R `fitdistrplus` package. The other "basic classical goodness-of-fit plot" is a plot of the fit vs empirical CDF. This and other plots can be added.

I agree that discrete samples below the PDF (continuous distributions only) may be better than a histogram if there are only a handful of samples. By default, we could switch between the two based on a point number threshold or some other visual clarity metric. I'd suggest that we leave these sorts of bells and whistles to future PRs. You could argue that the `plot` method itself is outside the scope of this PR, too, but I was willing to add something for that to pave the way for improvements.

Reply:
Meh, what do you think:
![image](https://user-images.githubusercontent.com/6570539/151281878-04f0a9a0-eef6-4eed-a344-167f8ac9803e.png)

vs

![image](https://user-images.githubusercontent.com/6570539/151282010-a5930732-7b53-47ac-88e2-f8bf835d6bc7.png)

I set the threshold at 50 points so we get histograms when there are too many points to see.